### PR TITLE
perf(inputs): keep inputs grid-sharded

### PR DESCRIPTION
**Pull request by japols**
## Description
Keep input and autoregressive rollout tensors sharded across model ranks, gathering forecast fields only to rank 0 when producing output.

Companion anemoi-models change: https://github.com/ecmwf/anemoi-core/pull/1345

## What problem does this change solve?
For high-resolution tasks like 4km downscaling, storing full non-grid-sharded inputs can take up a majority of GPU memory. This PR makes the memory usage for parallel inference scalable i.e. n gpus/model -> 1/n x memory requirements for storing input tensors.

> [!WARNING]
> Requires this anemoi-core change for compatibility: https://github.com/ecmwf/anemoi-core/pull/1345

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/Contributor-License-Agreement.md)

### CODE REVIEW

File: src/anemoi/inference/runner.py [hunk 1]
```diff
@@ -696,14 +711,17 @@ def execute(self) -> None:
...
 
         for output in self.outputs.values():
             output.close()
 
+    def write_output_state(self, dataset: str, state: State) -> None:
+        """Apply top-level post-processors and write one forecast state."""
+        for processor in self.post_processors[dataset]:
+            state = processor.process(state)
+        self.outputs[dataset].write_state(state)
+
```
Comment (HCookie) [inline src/anemoi/inference/runner.py:726]: This seems a change unrelated to this PR?

File: src/anemoi/inference/runners/parallel.py [hunk 2]
```diff
@@ -30,6 +30,28 @@
...
+    padded = torch.empty(padded_shape, dtype=tensor.dtype, device=tensor.device)
+    padded[: tensor.shape[0]].copy_(tensor)
+
+    gathered = [torch.empty_like(padded) for _ in shard_sizes] if is_master else None
+    torch.distributed.gather(padded, gather_list=gathered, dst=0, group=process_group)
+
+    if not is_master:
+        return None
+
+    return torch.cat([shard[:size] for shard, size in zip(gathered, shard_sizes)], dim=0)
```
Comment (HCookie) [inline src/anemoi/inference/runners/parallel.py:52]: This still materialises the full tensor on the master rank right? 
(just to check my understanding)

File: src/anemoi/inference/runner.py [hunk 3]
```diff
@@ -443,6 +457,7 @@ def forecast(
                 dataset: torch.from_numpy(np.swapaxes(input_tensor_numpy, -2, -1)[np.newaxis, ...]).to(self.device)
                 for dataset, input_tensor_numpy in input_tensors_numpy.items()
             }
+            input_tensors_torch = self.prepare_forecast_input_tensors(input_tensors_torch)
```
[the remark this answers is not in the document]
Comment (japols) [reply src/anemoi/inference/runner.py:460]: this is true but also a known limitation: afaik writing outputs in grib requires full fields (i.e. all gridpoints). a solution would be to either:
* chunk across the field/variable dimension (gather + write output in smaller chunks of fields)
* extend parallel output writing to all GPUs (instead of only spawning cpu writer tasks from rank 0) and perform an all-to-all operation to have full grids and sharded fields across GPUs that can then be written in parallel

thoughts @cathalobrien ? maybe good for a follow-up PR whenever needed

File: src/anemoi/inference/runners/parallel.py [hunk 4]
```diff
@@ -179,6 +202,103 @@ def seed(self, comm_group: "torch.distributed.ProcessGroup | None") -> None:
...
+        assert process_group is not None
+        rank = process_group.rank()
+        start = sum(shard_sizes[:rank])
+        return slice(start, start + shard_sizes[rank])
+
+    def prepare_forecast_output_state(self, state: dict[str, Any]) -> dict[str, Any]:
+        """Gather all forecast fields together on rank 0 while keeping rollout state sharded."""
+        process_group = self.compute_client.process_group
+        if process_group is None:
+            return state
```
[the remark this answers is not in the document]
Comment (japols) [reply src/anemoi/inference/runners/parallel.py:240]: follow-up PR potentially for @joherics to extend sharded tests to inference

File: src/anemoi/inference/runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (HCookie) [reply src/anemoi/inference/runner.py:460]: It is possible an integration with `multio` may assist with the sharded writing, but for now, yes writing outputs in grib requires full fields.
Although this comment seems related to the input not the output tensor, which also materialises the full grid.

File: src/anemoi/inference/runners/parallel.py (hunk 2 above)
Comment (japols) [reply src/anemoi/inference/runners/parallel.py:52]: yeah, but only after the model has been applied so memory pressure shouldn't be too high. also see my comment above on potential improvements

File: src/anemoi/inference/runner.py (hunk 1 above)
Comment (japols) [reply src/anemoi/inference/runner.py:726]: I wanted to unify the Runner API to overwrite write_output_state for the ParallelRunner (which calls super().write_output_state only on the master rank)

File: src/anemoi/inference/tensors.py [hunk 5]
```diff
@@ -499,10 +500,18 @@ def add_boundary_forcings_to_input_tensor(
             forcings = np.swapaxes(
                 forcings[np.newaxis, :, np.newaxis, ...], -2, -1
             )  # shape: (1, dates, 1, values, variables)
+
+            grid_slice = self.context.grid_shard_slice(self.dataset_name)
```
Comment (japols) [inline src/anemoi/inference/tensors.py:502]: @dietervdb-meteo could you review the sharding logic for the local spatial mask please?
